### PR TITLE
[Train] LightningTrainer converts relative ckpt dirpath to an absolute path.

### LINKED ISSUE #35894: [Train] LightningTrainer converts relative checkpoint dirpath to absolute path
### What happened + What you expected to happen

During restoration, PyTorch Lightning expects all workers to have the same directory structure. But when we specify a relative path for `dirpath`, PyTorch Lightning creates a new folder under the current working directory (the `rank_x` folder) like `.../LightningTrainer_7282d_00000_0_2023-05-26_01-43-54/rank_x/{dirpath}` for each worker. 

This will cause the internal state of the ModelCheckpoint callback to not be properly restored, resulting in inconsistent NCCL operations that resulted in timeouts. We need to find a proper way in LightningTrainer to convert relative checkpoint dirpaths to absolute paths to remove this issue.

### Versions / Dependencies

master

### Reproduction script

```python

import os
import torch
import torch.nn as nn
import torch.nn.functional as F
from filelock import FileLock
from torch.utils.data import DataLoader, random_split, Subset
from torchmetrics import Accuracy
from torchvision.datasets import MNIST
from torchvision import transforms

import pytorch_lightning as pl
from pytorch_lightning import trainer
from pytorch_lightning.core import datamodule
from pytorch_lightning.loggers.csv_logs import CSVLogger

class MNISTDataModule(pl.LightningDataModule):
    def __init__(self, batch_size=100):
        super().__init__()
        self.data_dir = os.getcwd()
        self.batch_size = batch_size
        self.transform = transforms.Compose(
            [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
        )

    def setup(self, stage=None):
        with FileLock(f"{self.data_dir}.lock"):
            mnist = MNIST(
                self.data_dir, train=True, download=True, transform=self.transform
            )

            
            self.mnist_train, self.mnist_val = random_split(mnist, [55000, 5000])

    def train_dataloader(self):
        return DataLoader(self.mnist_train, batch_size=self.batch_size, num_workers=4)

    def val_dataloader(self):
        return DataLoader(self.mnist_val, batch_size=self.batch_size, num_workers=4)

    def test_dataloader(self):
        with FileLock(f"{self.data_dir}.lock"):
            self.mnist_test = MNIST(
                self.data_dir, train=False, download=True, transform=self.transform
            )
        return DataLoader(self.mnist_test, batch_size=self.batch_size, num_workers=4)


datamodule = MNISTDataModule(batch_size=128)

class MNISTClassifier(pl.LightningModule):
    def __init__(self, lr=1e-3, feature_dim=128):
        torch.manual_seed(421)
        super(MNISTClassifier, self).__init__()
        self.linear_relu_stack = nn.Sequential(
            nn.Linear(28 * 28, feature_dim),
            nn.ReLU(),
            nn.Linear(feature_dim, 10),
            nn.ReLU(),
        )
        self.lr = lr
        self.accuracy = Accuracy(task="multiclass", num_classes=10)
        self.eval_loss = []
        self.eval_accuracy = []

    def forward(self, x):
        x = x.view(-1, 28 * 28)
        x = self.linear_relu_stack(x)
        return x

    def training_step(self, batch, batch_idx):
        if self.current_epoch == 3 and self.global_rank == 0:
            if not os.path.exists("/mnt/cluster_storage/error_flag"):
                os.system("touch /mnt/cluster_storage/error_flag")
                raise RuntimeError("Mock Error")

        import time
        time.sleep(0.01)
        x, y = batch
        y_hat = self(x)
        loss = torch.nn.functional.cross_entropy(y_hat, y)
        self.log("train_loss", loss)
        return loss

    def validation_step(self, val_batch, batch_idx):
        loss, acc = self._shared_eval(val_batch)
        self.log("val_accuracy", acc)
        self.eval_loss.append(loss)
        self.eval_accuracy.append(acc)
        return {"val_loss": loss, "val_accuracy": acc}

    def test_step(self, test_batch, batch_idx):
        loss, acc = self._shared_eval(test_batch)
        self.log("test_accuracy", acc)
        return {"test_loss": loss, "test_accuracy": acc}

    def _shared_eval(self, batch):
        x, y = batch
        logits = self.forward(x)
        loss = F.nll_loss(logits, y)
        acc = self.accuracy(logits, y)
        return loss, acc

    def on_validation_epoch_end(self):
        avg_loss = torch.stack(self.eval_loss).mean()
        avg_acc = torch.stack(self.eval_accuracy).mean()
        self.log("val_loss", avg_loss, sync_dist=True)
        self.log("val_accuracy", avg_acc, sync_dist=True)
        self.eval_loss.clear()
        self.eval_accuracy.clear()
    
        with open(f"/mnt/cluster_storage/ckpt_callback_{self.global_rank}.txt", 'w') as f:
            f.write(f"{self.trainer.checkpoint_callback.__dict__}")

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr)
        return optimizer

from pytorch_lightning.callbacks import ModelCheckpoint
from ray.air.config import FailureConfig, RunConfig, ScalingConfig, CheckpointConfig
from ray.train.lightning import (
    LightningTrainer,
    LightningConfigBuilder,
)

def build_lightning_config_from_existing_code(use_gpu):
    config_builder = LightningConfigBuilder()
    config_builder.module(cls=MNISTClassifier, lr=1e-3, feature_dim=128)
    config_builder.checkpointing(
        # dirpath=f'chkpts/{datetime.datetime.now().strftime("%d-%m-%y_%H:%M:%S")}/',
        dirpath="/tmp/my_ckpt_dir",
        monitor="val_accuracy", 
        mode="max", 
        save_top_k=3
    )
    config_builder.trainer(
        max_epochs=10000,
        accelerator="gpu" if use_gpu else "cpu",
        log_every_n_steps=100,
        logger=CSVLogger("logs"),
    )
    config_builder.fit_params(datamodule=datamodule)
    lightning_config = config_builder.build()
    return lightning_config


use_gpu = True
lightning_config = build_lightning_config_from_existing_code(use_gpu=use_gpu)
scaling_config = ScalingConfig(num_workers=2, use_gpu=use_gpu)

run_config = RunConfig(
    name="test-sync-artifacts",
    storage_path="/tmp/ray_results",
    checkpoint_config=CheckpointConfig(
        num_to_keep=3,
        checkpoint_score_attribute="val_accuracy",
        checkpoint_score_order="max",
    ),
    failure_config=FailureConfig(max_failures=-1),
)

trainer = LightningTrainer(
    lightning_config=lightning_config,
    scaling_config=scaling_config,
    run_config=run_config,
)

result = trainer.fit()
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

During restoration, PyTorch Lightning expects all workers to have the same directory structure. We need to make sure all workers have the same absolute dirpath for the ModelCheckpoint Callback.

When we specify a relative path for dirpath, PTL creates a dirpath in the working directory of each worker. Like `.../LightningTrainer_7282d_00000_0_2023-05-26_01-43-54/rank_x/{dirpath}`. This will cause the internal state of the ModelCheckpoint callback to not be properly restored, resulting in inconsistent NCCL operations that resulted in timeouts. We need to find a proper way in LightningTrainer to convert relative checkpoint dirpaths to absolute paths to remove this issue.

## Related issue number

Closes #35894 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/lightning/lightning_trainer.py [hunk 1]
```diff
@@ -527,6 +535,13 @@ def _lightning_train_loop_per_worker(config):
     trainer_config["enable_checkpointing"] = True
     model_checkpoint_config["save_last"] = True
 
+    # Convert dirpath to an absolute path under the trial directory
+    ckpt_dirpath = model_checkpoint_config.get("dirpath", None)
+    if ckpt_dirpath and not os.path.isabs(ckpt_dirpath):
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_trainer.py:540]: We should also expanduser, if the user passes in something like `"~/checkpoint_dirpath"`:

```python
os.path.isabs(os.path.expanduser(dirpath))
```

File: python/ray/train/lightning/lightning_trainer.py [hunk 2]
```diff
@@ -527,6 +535,13 @@ def _lightning_train_loop_per_worker(config):
...
+        model_checkpoint_config["dirpath"] = os.path.join(
+            session.get_trial_dir(), ckpt_dirpath
+        )
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_trainer.py:543]: One risk:
- Saving to the trial dir will upload the checkpoint again as an artifact. In the cloud checkpointing case.
- I think this is already the default behavior though. We should just make a tracker issue for this problem.

File: python/ray/train/tests/test_lightning_trainer.py [hunk 3]
```diff
@@ -71,11 +71,13 @@ def test_trainer_with_native_dataloader(
     batch_size = 8
     num_workers = 2
     dataset_size = 256
+    ckpt_dirpath = "lightning/ckpts/dir"
```
Comment (justinvyu) [inline python/ray/train/tests/test_lightning_trainer.py:74]: Should we test all the possibilities: default, with a relative ckpt_dirpath, and with an absolute ckpt_dirpath?

Also, let's add the dirpath to the tuning lightning release test.

File: python/ray/train/tests/test_lightning_trainer.py [hunk 4]
```diff
@@ -71,11 +71,13 @@ def test_trainer_with_native_dataloader(
...
 
     config_builder = (
         LightningConfigBuilder()
         .module(LinearModule, input_dim=32, output_dim=4, strategy=strategy)
         .trainer(max_epochs=num_epochs, accelerator=accelerator)
+        .checkpointing(dirpath=ckpt_dirpath)
```
Comment (justinvyu) [inline python/ray/train/tests/test_lightning_trainer.py:80]: Should we assert the contents of the trial folder after the run?

File: python/ray/train/lightning/lightning_trainer.py [hunk 5]
```diff
@@ -237,6 +238,13 @@ class LightningTrainer(TorchTrainer):
     using the arguments provided in ``LightningConfigBuilder.fit_params()`` and then
     run ``pytorch_lightning.Trainer.fit``.
 
+    For more flexible and fine-grained control over `LightningModule` and `pl.Trainer`,
+    you can specify one or more customized PyTorch Lightning Callbacks
+    in `LightningConfigBuilder.trainer(callbacks=)`.
+
+    LightningTrainer requires ``pytorch_lightning>=1.6.5`` package, and it is tested
+    with ``pytorch_lightning==1.6.5`` and ``pytorch_lightning==2.0.0``.
```
Comment (justinvyu) [inline python/ray/train/lightning/lightning_trainer.py:246]: Should these be captured by an import error? When importing ptl, check the version and raise an error if we try to use LightningTrainer with an incompatible version.

**Review by justinvyu — COMMENTED** [review]
Thanks, just a couple of questions:

File: python/ray/train/lightning/lightning_trainer.py (hunk 2 above)
Comment (woshiyyya) [reply python/ray/train/lightning/lightning_trainer.py:543]: Yeah I notice that we already upload the whole trial folder to the cloud storage. Maybe we can setup a flag to determine whether to upload the artifacts or not.

Comment (justinvyu) [reply python/ray/train/lightning/lightning_trainer.py:543]: There actually is a `sync_artifacts` flag that defaults to True. Can recommend that in the meantime.

File: python/ray/train/lightning/lightning_trainer.py (hunk 1 above)
Comment (woshiyyya) [reply python/ray/train/lightning/lightning_trainer.py:540]: Make sense : )

File: python/ray/train/lightning/lightning_trainer.py (hunk 5 above)
Comment (woshiyyya) [reply python/ray/train/lightning/lightning_trainer.py:246]: Some users complained that they don't know where to find the information about compatible ptl versions. So I put it here. What about we also add a import error to check the version compatibility.

File: python/ray/train/tests/test_lightning_trainer.py (hunk 3 above)
Comment (woshiyyya) [reply python/ray/train/tests/test_lightning_trainer.py:74]: Done

**Comment by woshiyyya** [comment]
Close this PR since we plan to chdir for all workers to `rank_all` folder, which is a more robust way to align with Lightning's settings.

**Pull request closed without merging**